src/Microsoft.OData.Core/ODataBatchWriter.cs

```diff
@@ -422,473 +388,307 @@ void IODataBatchOperationListener.BatchOperationContentStreamDisposed()
         /// This listener can choose to fail, if the currently written payload doesn't support in-stream error at this position.
         /// If the listener returns, the writer should not allow any more writing, since the in-stream error is the last thing in the payload.
         /// </remarks>
-        void IODataOutputInStreamErrorListener.OnInStreamError()
-        {
-            this.rawOutputContext.VerifyNotDisposed();
-            this.SetState(BatchWriterState.Error);
-            this.rawOutputContext.TextWriter.Flush();
-
-            // The OData protocol spec did not defined the behavior when an exception is encountered outside of a batch operation. The batch writer
-            // should not allow WriteError in this case. Note that WCF DS Server does serialize the error in XML format when it encounters one outside of a
-            // batch operation.
-            throw new ODataException(Strings.ODataBatchWriter_CannotWriteInStreamErrorForBatch);
-        }
+        public abstract void OnInStreamError();
 
         /// <summary>
-        /// Determines whether a given writer state is considered an error state.
-        /// </summary>
-        /// <param name="state">The writer state to check.</param>
-        /// <returns>True if the writer state is an error state; otherwise false.</returns>
-        private static bool IsErrorState(BatchWriterState state)
-        {
-            return state == BatchWriterState.Error;
-        }
-
-        /// <summary>
-        /// Verifies that calling WriteStartBatch is valid.
+        /// Verifies that calling CreateOperationRequestMessage is valid.
         /// </summary>
         /// <param name="synchronousCall">true if the call is to be synchronous; false otherwise.</param>
-        private void VerifyCanWriteStartBatch(bool synchronousCall)
+        /// <param name="method">The Http method to be used for this request operation.</param>
+        /// <param name="uri">The Uri to be used for this request operation.</param>
+        /// <param name="contentId">The Content-ID value to write in ChangeSet head.</param>
+        internal void CanCreateOperationRequestMessageVerifierCommon(bool synchronousCall, string method, Uri uri, string contentId)

```

Comment:
>CanCreateOperationRequestMessageVerifierCommon [](start = 22, length = 46)

By moving the logic in VerifyCanCreateOperationRequestMessageAgainstChangesetBoundary in the ODataMultipartMixedBatchReader to CreateOperationRequestMessageImplementation in ODataMultipartMixedBatchReader, you can move this logic into VerifyCanCreateOperationRequestMessage as a private method in the ODataBatchWriter base class, and get rid of this internal method.

Reply:
This would end up with verification logic in two places: inside VerifyCanCreateOperationRequestMessage; and inside CreateOperationRequestMessageImplementation (the VerifyCanCreateOperationRequestMessageAgainstChangesetBoundary). I think it breaks the logical, functional boundary between VerifyCanCreateOperationRequestMessage and its actual processing (CreateOperationRequestMessageImplementation) and therefore is not desirable?

---
In reply to: [140936780](https://github.com/OData/odata.net/pull/960#discussion_r140936780) [](ancestors = 140936780)

Reply:
Per our discussion offline, updated with a different approach keeping all verifications done prior to invoking Create...().

---
In reply to: [142585752](https://github.com/OData/odata.net/pull/960#discussion_r142585752) [](ancestors = 142585752,140936780)